### ISSUE TITLE
Looking for help refactoring the parsing code

### ISSUE THREAD

**Issue by david-janssen** [david-janssen/kmonad#237]
So, the parsing code is a bit of a mess. I'm creating this issue mainly because I had this on my list as something that eventually, after a list of other eventuallies, maybe I should have a look at.

Then I noticed that @slotThe did some great work in #152, adding `:keyword` options to or S-exprs. So I thought, maybe it would be a good idea to open an issue for this.

My thinking is:
- I like lisp as an easy to parse language
- Ideally we'd have things like simple, definable constants
- Keywords are a great idea, and could be used for other definitions as well (like adding option-configuration to layers).
- Macros would be really cool
- A giant goal would be to have something like MAL implemented in the config

Probably one of the first steps would be generalizing the parsing language, so that we have a module that turns raw kmonad-config stuff into S-exps with a head-term, arguments, potentially keywords and such.

And I was wondering if a certain person with a certain penchant for parsing felt inclined to have a go? If not, I'm just going to leave this open as a bin for people to dump their ideas about parsing and the config-language, that I will eventually sort through. If so, then it's the same kind of bin, but with a different benevolent dictator calling the shots.

**Comment by thoelze1** [comment]
I would be interested in taking this on. But first I want to clear up a bigger (perhaps out of the scope of this issue) question: at what point should a feature be implemented within kmonad as opposed to in the configuration file? Many a user (myself included) has come to this project with a specific idea that needs to be implemented across several Haskell files instead of in a single file of a configuration language with a lower barrier to entry.

But in theory, we could make a configuration language general enough that even complex buttons like `tap-next-release` could be implemented in a config. You [referred](https://github.com/kmonad/kmonad/issues/101#issuecomment-840358704) to this as "too big of a solution," and I somewhat agree. But I also think this would be really cool and fun to build :). Doing this would allow users with niche requests (#203, #222, etc) to easily (hopefully) implement their desired functionality in their config, while also allowing popular buttons to exist in the language or be available for import.

Maybe learning a lisp configuration language would itself be a significant barrier to entry. But with the right design, I don't think it would be more difficult than following through the current tutorial.

Something else to think about: recently I've been wishing that kmonad could become a general remapper not just of keyboard events to keyboard events but of input events (keyboard, mouse, pedal, etc) to input events. Perhaps this belongs as a different issue, but it's something to keep in mind if we decide to change the config language.

**Comment by david-janssen** [comment]
So my thinking was this:

### Why not a full lisp
Well, after we have parsed the config there is no reason for us to compute anything further, right? There is no 'end-goal' to calculate beyond a runnable config. I think we really should go no further than "remap input events to output events". Therefore, if we implement a full lisp into KMonad, what are we really gaining?

What we are doing is giving people a lot of rope to hang themselves with, and then come with increasingly complicated and weird issues. 

### What else?
Well, I've thought about this a lot, and I think it boils down to first asking ourselves this question:

#### What do we want to achieve?
My answer, I think, is something along the lines of:
- a language in which it is easy to specify matching certain input patterns in a stream of events to actions.
- an action language that is capable of expressing the generation of an output stream

I think there is an Algebra that we can specify here, although it's still hidden behind the mist of my non-comprehension at the moment. I was looking through "Algebra Driven Design" recently, and it's started giving me thoughts. Before I really start wrestling with these ideas I wanted to make sure that the core of KMonad was clean and solid, which is what I am investing my time in now, but once that was done I wanted to move on to this new model.

#### So what is really required?
I think if we can specify a language of very small but elegantly composing elements, then it should be straightforward for people to specify very arbitrary buttons. Just think about it: a tap-next-hold-release is actually just the following logic:
- notice my trigger being pressed
  - either
    - timeout -> do hold thing
    - notice my trigger being released -> do tap thing
    - notice something else being pressed
      - notice my trigger being released -> do tap thing
      - notice the something else being released -> do hold thing

And that is basically the most complicated button we have. I think if we can really dive down into the basic construction blocks of input-stream-matching-predicates, then we can come up with 3 or 4 very simple building blocks that we can use to build the wilderst buttons.

Once we have those blocks and some way of turning those blocks into actual KMonad actions, then wrapping these actions in some lispy syntax is super easy, and people could specify whatever they want in the config language.

So instead of implementing a really powerful language in the config, which will cause loads of people with different solutions and strange problems, and essentially turn is into Lisp developers. Instead, identify exactly the abstractions we need, wrap those in shallow lisp, and keep the logic in Haskell, which is a much nicer language for thinking.

**Comment by thoelze1** [comment]
> small but elegantly composing elements

> we can come up with 3 or 4 very simple building blocks that we can use to build the wilderst buttons

This is a great explanation. I like these ideas.

**Comment by slotThe** [comment]
This sounds great!  @thoelze1 feel free to reach out if you want to split the load at any point

**Comment by vherrmann** [comment]
I'd also like to contribute to a new config system, since this was the main reason I joined kmonad. ^^

**Comment by srithon** [comment]
Hello,
For the past couple months, I've been working on a side project called [KMonadX](https://github.com/srithon/kmonadx), which is a TOML-like extension language that transpiles to `kbd`. It's certainly not a replacement for the parsing code refactor, but I'm fairly happy with its current state and think it may be useful for other people as well. Currently, the features it offers are:
1. Declarative bindings
2. Layer-namespaced aliases
3. Access modifiers for aliases
4. Topographical sorting of aliases
5. Layer inheritance
6. Numeric and string constants with interpolation
7. Source layer derived from bindings

I have plans to add pseudo-functions like the following to further cut down configuration boilerplate, and more convenience features along the line.
```kbdx
create_homerow_mod = |key, modifier| (tap-hold-next-release 125 $key $modifier)
```

While the tool may not be for everyone, I hope it can hold over those who are eagerly awaiting the configuration language update. The syntax and overall approach is fairly opinionated, so there are bound to be some disagreements, but if you have any thoughts or critiques please feel free to let me know. In any case, this project was a great learning experience for me, so I won't be hurt :).

**Comment by jdevuyst** [comment]
First of all. I love the project and its potential!

Second. Given the name of the project I was surprised that the configuration format was not more programmable. I was hoping to use vim-style layers, but Kmonad currently doesn't have those, and so it looks like I may have to edit the Haskell code to make that happen. Now I certainly don't mind writing Haskell, but I think it would be conceptually more neat if we could build new features without recompiling Kmonad. 🙃

Since you (sort of) appear to be soliciting suggestions, here are two ideas that came to mind.

**Idea 1: Config files written in Dhall, using Elm-style event handling**

Concepts such as 'layers' and 'buttons' would be removed from the Haskell code. They would be reimplemented in Dhall. 

A Dhall config would be a value of type `EventHandler`, defined as follows (using Haskell syntax):

```haskell
type Keycode = Int

type TimerID = Int

type Timeout = Double

data Event = KeyPressed Keycode | KeyReleased Keycode | TimerFired TimerID

data Command = EmitKeyPress Keycode | EmitKeyRelease Keycode | RunCmd String | RegisterTimer TimerID Timeout

data EventHandler = forall state. EventHandler
  { initialState :: state
  , handleEvent  :: state -> Event -> (state, [Command])
  }
```

(It may be the case that Dhall doesn't have existentials, but you could work around that.)

This is basically Elm-style reactive programming and I think it might be a really nice way to configure a keyboard for us programmer types. And hopefully for non-programmers it would be quite usable too, provided that the right functions are handed to them to easily add layers, sticky keys, chords, and so on.

**Idea 2: State machines**

My main concern with the previous idea might be performance. If we care a lot about performance then perhaps we should have an intermediate language. I think state machines, with actions defined on either the transitions or nodes, would expressive enough. And evaluating those should be fast. We could guarantee that each event is processed in constant time (ignoring timeouts).

Basically the input would be a `DFA`:

```haskell
type Node = Int

data Event = KeyPressed Int | KeyReleased Int | Timeout Double

data Command = EmitKeyPress Int | EmitKeyRelease Int | RunCmd String

data Transition = Transition Event Node [Command]

data DFA = DFA
  { transitions :: Map Int [Transition]
  } -- By convention the DFA starts at node 0
```

I think it should be possible to compile the existing configuration format (including layers and buttons) to `DFA`.

If you wanted to generate DFAs from Haskell or Dhall you'd probably create a regular expressions DSL (with `Event` for tokens) that would then compile to a `DFA` value.

And, by the way, I'd love to help out with something like this!
